pkg/apply/types.go

```diff
@@ -0,0 +1,89 @@
+package apply
+
+// IntValueChanges stores changes in integer values (NumPartitions & ReplicationFactor)
+// if a topic is being created then Updated == Current
+type IntValueChanges struct {
+	Current int `json:"current"`
+	Updated int `json:"updated"`
+}
+
+// NewConfigEntry holds configs which are being added to a topic
+type NewConfigEntry struct {
+	Name  string `json:"name"`
+	Value string `json:"value"`
+}
+
+// ConfigEntryChanges holds configs to be updated, as well as their current and updated values
+type ConfigEntryChanges struct {
+	Name    string `json:"name"`
+	Current string `json:"current"`
+	Updated string `json:"updated"`
+}
+
+// ReplicaAssignmentChanges stores replica reassignment
+// if a topic is being created then UpdatedReplicas == CurrentReplicas
+// TODO: update Changes to actually support replica reassignment
+type ReplicaAssignmentChanges struct {
+	Partition       int   `json:"partition"`
+	CurrentReplicas []int `json:"currentReplicas"`
+	UpdatedReplicas []int `json:"updatedReplicas"`
+}
+
+// enum for possible Action values in ChangesTracker
+type ActionEnum string
+
+const (
+	ActionEnumCreate ActionEnum = "create"
+	ActionEnumUpdate ActionEnum = "update"
+)
+
+// NewChangesTracker stores the structure of a topic being created in an apply run
+// to eventually be printed to stdout as a JSON blob in subcmd/apply.go
+type NewChangesTracker struct {
+	// Action records whether this is a topic being created or updated
+	Action            ActionEnum        `json:"action"`
+	Topic             string            `json:"topic"`
+	NumPartitions     int               `json:"numPartitions"`
+	ReplicationFactor int               `json:"replicationFactor"`
+	ConfigEntries     *[]NewConfigEntry `json:"configEntries"`
+}
+
+// UpdateChangesTracker stores the same data as NewChangesTracker,
+// but
+// to eventually be printed to stdout as a JSON blob in subcmd/apply.go
+type UpdateChangesTracker struct {
+	// Action records whether this is a topic being created or updated
+	Action ActionEnum `json:"action"`
```

Comment:
Why do we need an action field if the type (`UpdateChangesTracker` / `NewChangesTracker`) already defines the action ?

Reply:
we don't, I just find it more convenient to check `topic[action]` when working with a single change object than pull the key from a higher-level dict in my python script

it's redundant but I don't really see the harm in keeping it around, I can delete if you don't like it tho